services/src/main/java/io/scalecube/services/inject/ScaleCubeServicesProvider.java

```diff
@@ -0,0 +1,111 @@
+package io.scalecube.services.inject;
+
+import io.scalecube.services.Microservices;
+import io.scalecube.services.ServiceDefinition;
+import io.scalecube.services.ServiceInfo;
+import io.scalecube.services.ServiceProvider;
+import io.scalecube.services.ServicesProvider;
+
+import java.util.Collection;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import reactor.core.publisher.Mono;
+
+/**
+ * Default {@link ServicesProvider}. Thread-safe.
+ */
+public class ScaleCubeServicesProvider implements ServicesProvider {
+
+  private final Function<Microservices, Collection<ServiceInfo>> serviceFactory;
+
+  // lazy init
+  private AtomicReference<Collection<ServiceInfo>> services = new AtomicReference<>();
```

Comment:
Remove it at all

Reply:
> Remove it at all, see suggestions

Where those suggestions are?

Reply:
@segabriel I'm sorry, I didn't really know exactly how to do this.